File: Lab-FInal/1.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

#define NUM_STUDENTS 10
#define NUM_SLICES_PER_PIZZA 8


sem_t pizza;        
sem_t deliver;      
int num_slices=0; 
pthread_mutex_t mutex; 


void *student_thread(void *arg) 
{
    int student_id=*((int *)arg);
    while (1) 
    {
        sem_wait(&pizza); 
        pthread_mutex_lock(&mutex);
        if(num_slices>0) 
        {
            --num_slices; 
            printf("Student %d is eating pizza. Slices left: %d\n",student_id,num_slices);
        } 
        else 
        {
            printf("Student %d found no pizza. Going to sleep.\n",student_id);
            sem_post(&deliver);
            pthread_mutex_unlock(&mutex);
            sem_wait(&pizza); 
            pthread_mutex_lock(&mutex);
            if(num_slices>0)
            {
                --num_slices; 
                printf("Student %d is eating pizza. Slices left: %d\n",student_id,num_slices);
            }           
        }
        pthread_mutex_unlock(&mutex);
        sleep(1);
    }
    return NULL;
}

void *delivery_thread(void *arg) 
{
    while(1) 
    {
        sem_wait(&deliver); 
        pthread_mutex_lock(&mutex);
        if (num_slices==0) 
        {
            num_slices+=NUM_SLICES_PER_PIZZA; 
            printf("Pizza delivered. Slices left: %d\n",num_slices);
            pthread_mutex_unlock(&mutex);
            for(int i=0;i<NUM_STUDENTS;++i)
            {
                sem_post(&pizza);
            }
        } 
        else 
        {
            pthread_mutex_unlock(&mutex);
        }
    }
    return NULL;
}

int main() {
    pthread_t students[NUM_STUDENTS];
    pthread_t delivery;
    int student_ids[NUM_STUDENTS];
    
    sem_init(&pizza,0,0);
    sem_init(&deliver,0,1);
    pthread_mutex_init(&mutex,NULL);
    

    for(int i=0;i<NUM_STUDENTS;++i) 
    {
        student_ids[i]=i+1;
        pthread_create(&students[i],NULL,student_thread,&student_ids[i]);
    }
    
    pthread_create(&delivery,NULL,delivery_thread,NULL);
    

    for(int i=0;i<NUM_STUDENTS;++i) 
    {
        pthread_join(students[i],NULL);
    }
    pthread_join(delivery,NULL);
    
    sem_destroy(&pizza);
    sem_destroy(&deliver);
    pthread_mutex_destroy(&mutex);
    
    return 0;
}

File: Lab-FInal/1.py
import threading
import time

num_slices = 0
mutex = threading.Lock()
pizza = threading.Semaphore(0)
deliver = threading.Semaphore(1)

def student_thread(student_id):
    global num_slices
    while True:
        pizza.acquire()  
        mutex.acquire()
        if num_slices>0:
            num_slices-=1  
            print(f"Student {student_id} is eating pizza. Slices left: {num_slices}")
        else:
            print(f"Student {student_id} found no pizza. Going to sleep.")
            deliver.release()
            mutex.release()  
            pizza.acquire()  
            mutex.acquire()
            if num_slices>0:
                num_slices-=1  
                print(f"Student {student_id} is eating pizza. Slices left: {num_slices}")
            else:
                print(f"No pizza available even after delivery. Going to sleep.")
        mutex.release()
        time.sleep(1)


def delivery_thread():
    global num_slices
    while True:
        deliver.acquire()  
        mutex.acquire()
        if num_slices==0:
            num_slices+=8  
            print(f"Pizza delivered. Slices left: {num_slices}")
            mutex.release()
            for _ in range(num_students):  
                pizza.release()
        else:
            mutex.release()


num_students=10

student_threads=[]
for i in range(num_students):
    t=threading.Thread(target=student_thread,args=(i+1,))
    student_threads.append(t)
    t.start()

delivery_thread=threading.Thread(target=delivery_thread)
delivery_thread.start()

# Join threads
for t in student_threads:
    t.join()
delivery_thread.join()
